src/routes/solid-start/guides/data-apis.mdx

```diff
@@ -0,0 +1,596 @@
+
+```tsx {1, 4-14} 
+import { action } from "@solidjs/router"
+import { storage } from "~/lib/db"
+
+const addTodo = action(async (formData: FormData) => {
+  "use server"
+
+  const title = formData.get("title")
+
+  const todos = await storage.getItem("todos:data")
+
+  const newTodo = { id: uuid(), title, completed: false }
+
+  await storage.setItem("todos:data", [ ...todos, newTodo ])
+}, "add-todo")
+```
+
+Similar to queries, actions can also be server functions with direct access to server resources like database and external APIs.
+
+### Invoking actions
+
+There are two ways to invoke an action.
+
+#### Form Actions
+
+An action can be directly passed to the `action` property of a `<form>` tag. The action will receive a web standard `FormData` object as its input.
+
+```tsx {4, 18} 
+import { action } from "@solidjs/router"
+import { storage } from "~/lib/db"
+
+const addTodo = action(async (formData: FormData) => {
+  "use server"
+
+  const title = formData.get("title")
+
+  const todos = await storage.getItem("todos:data")
+
+  const newTodo = { id: uuid(), title, completed: false }
+
+  await storage.setItem("todos:data", [ ...todos, newTodo ])
+}, "add-todo")
+
+function AddTodo() {
+  return (
+    <form action={addTodo} method="post">
+      <input
+        name="title"
+        placeholder="What needs to be done?"
+        autofocus
+      />
+    </form>
+  )
+}
+```
+
+This approach supports **progressive enhancement**, i.e., works with JavaScript turned off, provided that the form was rendered on the server. You can also add an `onSubmit` handler for client-side logic when JavaScript is turned on, like validation or clearing the inputs.
+
+```tsx {7, 12-15, 20} 
+import { action } from "@solidjs/router"
+import { storage } from "~/lib/db"
+
+const addTodo = action(async (formData: FormData) => {...}, "add-todo")
+
+function AddTodo() {
+  let inputRef!: HTMLInputElement
+  return (
+    <form
+      action={addTodo}
+      method="post"
+      onSubmit={(e) => {
+        if (!inputRef.value.trim()) e.preventDefault()
+        setTimeout(() => (inputRef.value = ""))
+      }}
+    >
+      <input
+        name="title"
+        placeholder="What needs to be done?"
+        ref={inputRef}
+        autofocus
+      />
+    </form>
+  )
+}
+```
+Calling `e.preventDefault()` will prevent running the action.
+
+You can find more information about invoking actions with progressive enhancement [here](/solid-router/reference/data-apis/action).
+{/* 
+Actions can be bound with an additional value using `.with()`. This value becomes the first argument to the action, and the form data becomes the second. This is a convenient way to pass additional values into the action without resorting to creating hidden inputs manually.
+
+```tsx {4, 22} 
+import { action } from "@solidjs/router"
+import { storage } from "~/lib/db"
+
+const editTodo = action(async (id: number, formData: FormData) => {
```

Comment:
does `with` still pass the formData as the second argument? If so thats pretty handy and didn't know they could be combined like that

Reply:
yep this example is straight from todomvc